Support whitelisting dirs

ericfranz opened the pull request:
There are definitely cleaner ways to do this, or more robust ways.

Two UX fixes:

- [ ] 1. Try to download a zip of the selected directory that is not in the whitelist, and a 403 error occurs but a "Forbidden" error alert doesn't display like in the other Ajax requests that get a 403.
- [ ] 2. If doing a non AJAX request (like enter the forbidden URL into the browser) instead of the page showing "forbidden" it should probably present a list of links the user can access (the whitelisted directories) or go to home directory (or root directory that is set for the tree).

A more robust solution would be to add checks at the source - the actual code that hits the filesystem could raise the same exceptions that occur when a file or directory is actually not readable by a user. But that would take more time than this.

ericfranz commented:
Slightly cleaner might be a whitelist object:

```js
whitelist = {
    paths:  process.env.WHITELIST ? process.env.WHITELIST.split(":") : [],
    enabled: ( ) => { return this.paths.length > 0; },
    contains: (path) => {
        // whitelist path contains function here
    }
}

// then 
if(whitelist.enabled){
    app.use( //...

      if(whitelist.contains(filepath))
          next();
      else
          // error
```
```

ericfranz commented:
This should also change. Instead of blacklisting the request urls that contain file system paths outside the list of whitelisted paths, I should explicitly specify all the request urls that are whitelist, and black list everything else. That way if we miss a URL it is just denied.

ericfranz commented on app.js:
```diff
@@ -38,6 +60,59 @@ socket = io.listen(server, {
+        if(request_url == "/" || request_url == "/fs"){
+            filepath = "/";
+        }
+        else if(filepath != null && whitelist.contains(filepath)) {
```
separate if block

```js
if(request_url == "/" || request_url == "/fs"){
  filepath = "/";
}


if(filepath != null && whitelist.contains(filepath)) {
```

ericfranz commented on app.js:
```diff
@@ -22,13 +23,34 @@ var http        = require('http'),
+    paths:  process.env.WHITELIST ? process.env.WHITELIST.split(":") : [],
+    enabled: function(){ return this.paths.length > 0; },
+    contains: function(filepath){
+        return this.paths.filter(function(whitelisted_path){
```
This method should call `path.realpath(filepath)` (and handle the exception case etc.) to avoid filepath being a symlink, prior to testing against whitelist

MorganRodgers commented on app.js:
```diff
@@ -136,6 +212,7 @@ app.use(function (req, res, next) {
 try {
     app_version = gitSync.tag();
 } catch(error) {
+    # FIXME: await file read VERSION if it exists
```
Are you intentionally breaking here? This is not a valid JS comment.

MorganRodgers commented on app.js:
```diff
@@ -38,6 +60,59 @@ socket = io.listen(server, {
+//
+// url: /fs/Applications/
+// path: /Applications/
+if(whitelist.enabled()) {
```
From this context `whitelist.enabled()` is never `true`. Instead of the middleware being conditionally defined we should have an early exit inside the lambda.

ericfranz replied:
That is by design. `whitelist.enabled()` is true only when WHITELIST env is set. Oops, its supposed to be WHITELIST_PATH. Will fix that.

MorganRodgers commented:
Branch [whitelist_nodejs6_morgans ](https://github.com/OSC/ood-fileexplorer/tree/whitelist_nodejs6_morgans) fixes a few things including the not-a-js-comment, the middleware never getting defined (and thus not used), and adds both realpath and tilda expansion.

ericfranz commented:
@MorganRodgers I opened your branch as a PR to merge into this branch: https://github.com/OSC/ood-fileexplorer/pull/183

I also branched off of yours and added a PR addressing all the issues I mentioned in my previous comment except for the tilde expansion issue, which IMHO can wait. https://github.com/OSC/ood-fileexplorer/pull/184 is set to merge into #183 which can merge into this #182
